solarforecastarbiter/validation/validator.py

```diff
@@ -0,0 +1,281 @@
+            {'zenith_bounds': [75, 93], 'ghi_bounds': [50, np.Inf],
+             'ratio_bounds': [0.0, 1.10]}}}
+
+
+def _check_limits(val, lb=None, ub=None, lb_ge=False, ub_le=False):
+    if lb_ge:
+        lb_op = np.greater_equal
+    else:
+        lb_op = np.greater
+    if ub_le:
+        ub_op = np.less_equal
+    else:
+        ub_op = np.less
+
+    if (lb is not None) & (ub is not None):
+        return lb_op(val, lb) & ub_op(val, ub)
+    elif lb is not None:
+        return lb_op(val, lb)
+    elif ub is not None:
+        return ub_op(val, ub)
+    else:
+        raise ValueError('must provide either upper or lower bound')
+
+
+def _QCRad_ub(dni_extra, sza, coeff):
+    return coeff['mult'] * dni_extra * cosd(sza)**coeff['exp'] + coeff['min']
+
+
+def check_irradiance_limits_QCRad(irrad, test_dhi=False, test_dni=False,
+                                  coeff=None):
+    """
+    Tests for physical limits on GHI using the QCRad criteria.
+
+    Also tests for physical limits on DHI and DNI if test_dhi and test_dni,
+    respectively, are set and these data are present.
+
+    Test passes if a value > lower bound and value < upper bound. Lower bounds
+    are constant for all tests. Upper bounds are calculated as
+
+    .. math::
+        ub = min + mult * dni_extra * cos( solar_zenith)^exp
+
+    Parameters:
+    -----------
+    irrad : DataFrame
+        ghi : float
+            Global horizontal irradiance in W/m^2
+        solar_zenith : float
+            Solar zenith angle in degrees
+        dni_extra : float
+            Extraterrestrial normal irradiance in W/m^2
+        dhi : float, optional
+            Diffuse horizontal irradiance in W/m^2
+        dni : float, optional
+            Direct normal irradiance in W/m^2
+    test_dhi : boolean, default False
+    test_dni : boolean, default False
+    coeff : dict, default QCRAD_LIMITS
+        for keys 'ghi_ub', 'dhi_ub', 'dni_ub', value is a dict with
+        keys {'mult', 'exp', 'min'}. For keys 'ghi_lb', 'dhi_lb', 'dni_lb',
+        value is a float.
+
+    Returns:
+    --------
+    flags : DataFrame
+        physical_limit_flag : boolean
+            True if value passes physically-possible test
+        climate_limit_flag : boolean
+            True if value passes climatological test
+    """
+
+    if not coeff:
+        coeff = QCRAD_LIMITS
+
+    flags = pd.DataFrame(index=irrad.index, data=None,
+                         columns=['ghi_physical_limit_flag'])
+
+    try:
+        ghi_ub = _QCRad_ub(irrad['dni_extra'], irrad['solar_zenith'],
+                           coeff['ghi_ub'])
+    except KeyError:
+        raise KeyError('Requires solar_zenith and dni_extra')
+
+    try:
+        flags['ghi_physical_limit_flag'] = _check_limits(irrad['ghi'],
+                                                         coeff['ghi_lb'],
+                                                         ghi_ub)
+    except KeyError:
+        raise KeyError('ghi not found')
```

Comment:
Should we be concerned with the case where a `coeff` is supplied that doesn't contain `ghi_lb`, leading to a `KeyError('ghi not found')`? Same applies to the other try blocks. Maybe we should allow the original except to be raised rather than catching it and reraising a new one. Up for debate.

Reply:
How about `if ghi in irrad.keys`, then `KeyError` would be raised for defective `coeff` only.  Testing for a valid `coeff` isn't all that appealing.

Reply:
Second thought, I should write a helper to validate that `coeff`, since `coeff` is assumed to be valid in multiple places.

Reply:
I'm not crazy about naming that variable `coeff` got a better idea?

Reply:
Personally I don't think it's necessary to catch these `KeyErrors` because it's fairly safe to assume that all users of the python functions will understand their meaning. To be clear, these messages would not propagate to web dashboard or API users. 

Maybe `limits` instead of `coeff`?

Reply:
`limits` is better.  I can take the KeyError out, its de factor input checking.  Would make the code easier to read.